### PR TITLE
feat(tagging): Select which upstream stages to include in image search

### PR DESCRIPTION
**Pull request by jervi**
When tagging an image, the default behaviour is to look for images in all ancestor stages. In some cases, this is not the desired behaviour. This commit makes it possible to optionally define a list of stages that will be considered in the tagging stage. ~Will be accompanied with a change in Deck.~ Corresponds with the following Deck PR: https://github.com/spinnaker/deck/pull/5145

### CODE REVIEW

**Comment by jervi** [comment]
@spinnaker/reviewers PTAL

**Comment by anotherchrisberry** [comment]
(similar comment made in the deck PR)

This should work off stage `refId` fields instead of `name`s.

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTaggerSpec.groovy [hunk 1]
```diff
@@ -35,33 +35,41 @@ abstract class ImageTaggerSpec<T extends ImageTagger> extends Specification {
...
-    def stage3 = new Stage(pipeline, "stage3", [:])
+    def pipeline = new Execution(Execution.ExecutionType.PIPELINE, "orca")
+    def stage1 = new Stage(pipeline, "bake", "stage1", stage1Context + [cloudProvider: cloudProvider])
+    def stage2 = new Stage(pipeline, "findImageFromTags", "stage2", stage2Context + [cloudProviderType: cloudProvider])
+    def stage3 = new Stage(pipeline, "bake", "stage3", [:])
 
-    stage3.requisiteStageRefIds = [stage2.id.toString()]
-    stage2.requisiteStageRefIds = [stage1.id.toString()]
+    stage3.requisiteStageRefIds = [stage2.name]
+    stage2.requisiteStageRefIds = [stage1.name]
```
Comment (anotherchrisberry) [inline orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTaggerSpec.groovy]: Can we change these assignments to `[stageX.refId]`? Although technically allowed, it's a little confusing to have the `name` and `refId` be the same value, so maybe move these two assignments below the next block where the `refId`s are being added to the stages.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java [hunk 2]
```diff
@@ -96,12 +105,24 @@ protected Collection findImages(Collection<String> imageNames, Stage stage, Clas
   }
 
   @VisibleForTesting
-  Collection<String> upstreamImageIds(Stage sourceStage, String cloudProviderType) {
-    Collection<Stage> imageProvidingAncestorStages = sourceStage.ancestors().stream().filter((Stage stage) -> {
+  Collection<String> upstreamImageIds(Stage sourceStage, Collection<String> consideredStages, String cloudProviderType) {
+    List<Stage> ancestors = sourceStage.ancestors();
+    Collection<Stage> imageProvidingAncestorStages = ancestors.stream().filter((Stage stage) -> {
       String cloudProvider = (String) stage.getContext().getOrDefault("cloudProvider", stage.getContext().get("cloudProviderType"));
```
Comment (anotherchrisberry) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java]: Could you consolidate the logic below into this filter? Something like:
```java
boolean consideredStageRefIdMatches = consideredStages == null || consideredStages.isEmpty() || consideredStages.contains(stage.refId);
(stage.getContext().containsKey("imageId") || stage.getContext().containsKey("amiDetails")) && cloudProviderType.equals(cloudProvider) && consideredStageRefIdMatches
```
Or is there more subtle complexity in the collection of parent refIds below that makes it easier to keep as separate methods?

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTaggerSpec.groovy (hunk 1 above)
Comment (jervi) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTaggerSpec.groovy]: Fixed

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java [hunk 3]
```diff
@@ -44,7 +53,7 @@
   /**
    * @return true when, according to the underlying cloud provider, image tags have been updated to match the respective target
    */
-  abstract protected boolean areImagesTagged(Collection<Image> targetImages, Stage stage);
+  abstract protected boolean areImagesTagged(Collection<Image> targetImages, Collection<String> consideredStages, Stage stage);
```
[the remark this answers is not in the document]
Comment (jervi) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java]: Fixed

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java [hunk 4]
```diff
@@ -96,12 +105,24 @@ protected Collection findImages(Collection<String> imageNames, Stage stage, Clas
   }
 
   @VisibleForTesting
-  Collection<String> upstreamImageIds(Stage sourceStage, String cloudProviderType) {
-    Collection<Stage> imageProvidingAncestorStages = sourceStage.ancestors().stream().filter((Stage stage) -> {
+  Collection<String> upstreamImageIds(Stage sourceStage, Collection<String> consideredStages, String cloudProviderType) {
```
[the remark this answers is not in the document]
Comment (jervi) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java]: Fixed

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java (hunk 2 above)
Comment (jervi) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java]: Yeah, this was overly complicated. I've simplified it a lot using your suggestion. Thanks.

**Comment by robzienert** [comment]
@jervi @jrsquared -- Presumably since the Deck PR looks good, we need to get this one in first?

**Comment by jrsquared** [comment]
Presumably.

**Comment by jervi** [comment]
That is the preferred order, yes. 😃
But nothing will break if Deck for some reason is deployed before Orca.

**Pull request merged**